lib/linked_list.rb

```diff
@@ -12,150 +12,265 @@ def initialize(value, next_node = nil)
+  def delete(value)
+    return if !@head
+    current = @head
+    previous = nil
+
+    while current
+      if current.data == value
+        @head = current.next if current == @head
+        previous.next = current.next if previous # relink list if deleted value is in the middle
+        current = nil
+        return
+      else 
+        previous = current
+        current = current.next     
+      end
     end
+    
+    return nil
+  end
 
+  # method to reverse the singly linked list
+  # note: the nodes should be moved and not just the values in the nodes
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def reverse
+    return if !@head
+    current = @head
+    previous = nil
 
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_first
-      raise NotImplementedError
+    # for each node, reassign @next to the previous node
+    while current
+      @head = current if !current.next # reassign head if we reach the end of the list
+      temp = previous
+      previous = current
+      current = current.next
+      previous.next = temp
     end
 
-    # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def add_last(value)
-      raise NotImplementedError
-    end
+    return nil
+  end
 
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def get_last
-      raise NotImplementedError
-    end
 
-    # method to insert a new node with specific data value, assuming the linked
-    # list is sorted in ascending order
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def insert_ascending(value)
-      raise NotImplementedError
+  ## Advanced Exercises
+  # returns the value at the middle element in the singly linked list
+  # Time Complexity: O(n), where n is the number of nodes in the list (the time complexity of length and find_nth_from_end are both O(n))
+  # Space Complexity: O(1)
+  def find_middle_value
+    return find_nth_from_end(length() / 2)
+  end
+
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    last_index = length() - 1
+    return get_at_index(last_index - n)
+  end
+
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity: O(n), where n is the number of nodes in the list
+  # Space Complexity: O(n), a new array is created whose size is dependent on the number of nodes in the list
+  def has_cycle

```

Comment:
The time complexity is O(n^2) because `.include` searches the entire array for an instance of the node.

There is a faster way with 2 node points one which move two hops each iteration and the other move one hop.